pkg/kubectl/pdb.go

```diff
@@ -95,7 +102,119 @@ func (s *PodDisruptionBudgetV1Generator) validate() error {
 		return fmt.Errorf("a selector must be specified")
 	}
 	if len(s.MinAvailable) == 0 {
-		return fmt.Errorf("the minimim number of available pods required must be specified")
+		return fmt.Errorf("the minimum number of available pods required must be specified")
+	}
+	return nil
+}
+
+// PodDisruptionBudgetV2Generator supports stable generation of a pod disruption budget.
+type PodDisruptionBudgetV2Generator struct {
+	Name           string
+	MinAvailable   string
+	MaxUnavailable string
+	Selector       string
+}
+
+// Ensure it supports the generator pattern that uses parameters specified during construction.
+var _ StructuredGenerator = &PodDisruptionBudgetV2Generator{}
+
+func (PodDisruptionBudgetV2Generator) ParamNames() []GeneratorParam {
+	return []GeneratorParam{
+		{"name", true},
+		{"min-available", false},
+		{"max-unavailable", false},
+		{"selector", false},
+	}
+}
+
+func (s PodDisruptionBudgetV2Generator) Generate(params map[string]interface{}) (runtime.Object, error) {
+	err := ValidateParams(s.ParamNames(), params)
+	if err != nil {
+		return nil, err
+	}
+
+	name, isString := params["name"].(string)
+	if !isString {
+		return nil, fmt.Errorf("expected string, saw %v for 'name'", name)
+	}
+
+	minAvailable, isString := params["min-available"].(string)
+	if !isString {
+		return nil, fmt.Errorf("expected string, found %v", minAvailable)
+	}
+
+	maxUnavailable, isString := params["max-available"].(string)
+	if !isString {
+		return nil, fmt.Errorf("expected string, found %v", maxUnavailable)
+	}
+
+	selector, isString := params["selector"].(string)
+	if !isString {
+		return nil, fmt.Errorf("expected string, found %v", selector)
+	}
+	delegate := &PodDisruptionBudgetV2Generator{Name: name, MinAvailable: minAvailable, MaxUnavailable: maxUnavailable, Selector: selector}
+	return delegate.StructuredGenerate()
+}
+
+// StructuredGenerate outputs a pod disruption budget object using the configured fields.
+func (s *PodDisruptionBudgetV2Generator) StructuredGenerate() (runtime.Object, error) {
+	if err := s.validate(); err != nil {
+		return nil, err
+	}
+
+	selector, err := metav1.ParseToLabelSelector(s.Selector)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(s.MaxUnavailable) == 0 && len(s.MinAvailable) == 0 {
+		s.MinAvailable = "1"
+
+		// This behavior is intended for backward compatibility.
+		// TODO: remove in Kubernetes 1.8
+		fmt.Fprintln(os.Stderr, "Deprecated behavior in kubectl create pdb: Defaulting min-available to 1. "+
+			"Kubernetes 1.8 will remove this default, and one of min-available/max-available must be specified. ")
```

Comment:
How about saying that we will default to maxUnavailable? Are we going to support both options long-term or is it that MaxUnavailable supersedes MinAvailable?